### PR TITLE
Build Prow Image from Scratch

### LINKED ISSUE #2957: Build the prow-tests image from scratch instead of from the kubekins-e2e image
The current prow-tests image we use for running all Prow jobs is based on the kubekins-e2e image - see https://github.com/knative/test-infra/blob/main/images/prow-tests/Dockerfile#L15, which is large and contains a lot of tools that are not needed by Knative. The latest one of gcr.io/knative-tests/test-infra/prow-tests is 2.9 GB.

To reduce its size and make it more clear on the version control, it'll be more ideal to build the image from scratch and only install the tools that are needed. We can learn from how Istio built their images - https://github.com/istio/tools/blob/master/docker/build-tools/Dockerfile

### PR DESCRIPTION
**Pull request by upodroid**
**What this PR does, why we need it**:

The upstream prow image from kubernetes/test-infra is deprecated so we need to build our image from scratch.

**Which issue(s) this PR fixes**:
Fixes #2957
Closes: #3012

**Special notes to reviewers**:

**User-visible changes in this PR**:

### CODE REVIEW

**Comment by upodroid** [comment]
/assign @chizhg 

Did I miss any tools?

Also, how do you want to deal with the missing in runner.sh?

```
REDACTED  MCW0CDP3YY  ~  Desktop  …  kn-test-infra  images  prow-tests   bump-maven  1✎  1⚑  USAGE  $    make build
docker build --no-cache --pull -t gcr.io/knative-tests/test-infra/prow-tests:v20220114-08024dc4-dirty -f Dockerfile  ../..
[+] Building 163.4s (27/52)                                                                                                        
 => [internal] load build definition from Dockerfile                                                                          0.1s
 => => transferring dockerfile: 8.57kB                                                                                        0.0s
 => [internal] load .dockerignore                                                                                             0.0s
 => => transferring context: 2B                                                                                               0.0s
 => [internal] load metadata for docker.io/library/golang:1.17                                                                3.0s
 => [internal] load metadata for docker.io/google/cloud-sdk:368.0.0-slim                                                      3.0s
 => [internal] load build context                                                                                             1.7s
 => => transferring context: 251.23kB                                                                                         1.5s
 => CACHED [stable-stuff  1/21] FROM docker.io/google/cloud-sdk:368.0.0-slim@sha256:26017397f4c2b8ef69fe6f8f019984007527c13c  0.0s
 => CACHED [external-go-gets  1/17] FROM docker.io/library/golang:1.17@sha256:0fa6504d3f1613f554c42131b8bf2dd1b2346fb69c2fc2  0.0s
 => [external-go-gets  2/17] RUN go install github.com/google/ko@v0.8.2                                                     134.2s
 => [stable-stuff  2/21] RUN curl -fsSL "https://storage.googleapis.com/kubernetes-release/release/v1.21.4/bin/linux/amd64/k  6.2s
 => [stable-stuff  3/21] RUN curl -fsSL "https://github.com/bazelbuild/bazelisk/releases/download/v1.11.0/bazelisk-linux-amd  2.5s
 => [stable-stuff  4/21] RUN curl -fsSL https://download.docker.com/linux/ubuntu/gpg | apt-key add - &&     echo "deb [arch  35.2s
 => [stable-stuff  5/21] RUN curl -sf https://adoptopenjdk.jfrog.io/adoptopenjdk/api/gpg/key/public | apt-key add -           6.1s
 => [stable-stuff  6/21] RUN echo "deb https://adoptopenjdk.jfrog.io/adoptopenjdk/deb buster main" | tee /etc/apt/sources.li  0.6s
 => [stable-stuff  7/21] RUN apt-get update                                                                                   4.0s
 => [stable-stuff  8/21] RUN apt-get install -y adoptopenjdk-16-hotspot                                                      44.2s
 => [stable-stuff  9/21] RUN curl -fsSL https://archive.apache.org/dist/maven/maven-3/3.8.4/binaries/apache-maven-3.8.4-bin.  1.7s
 => [stable-stuff 10/21] RUN java -version &&     mvn -v                                                                      1.4s
 => [stable-stuff 11/21] RUN apt-get install -y bison uuid-runtime shellcheck unzip                                           5.5s
 => [stable-stuff 12/21] RUN curl -s -S -L https://raw.githubusercontent.com/moovweb/gvm/master/binscripts/gvm-installer | G  5.6s
 => [stable-stuff 13/21] COPY images/prow-tests/source-gvm-and-run.sh /usr/local/bin                                          0.1s
 => [stable-stuff 14/21] RUN source-gvm-and-run.sh install go1.15.15 --prefer-binary                                         15.0s
 => [stable-stuff 15/21] RUN source-gvm-and-run.sh install go1.16.9 --prefer-binary                                          12.4s
 => CANCELED [external-go-gets  3/17] RUN go get github.com/boz/kail/cmd/kail                                                25.8s
 => [stable-stuff 16/21] RUN source-gvm-and-run.sh install go1.17.2 --prefer-binary                                          15.0s
 => [stable-stuff 17/21] RUN source-gvm-and-run.sh use go1.17 --default                                                       1.5s
 => [stable-stuff 18/21] RUN curl -fsSL "https://github.com/protocolbuffers/protobuf/releases/download/v3.17.0/protoc-3.17.0  1.5s
 => ERROR [stable-stuff 19/21] RUN mv "$(which runner.sh)" "$(dirname "$(which runner.sh)")/kubekins-runner.sh"               0.6s
------
 > [stable-stuff 19/21] RUN mv "$(which runner.sh)" "$(dirname "$(which runner.sh)")/kubekins-runner.sh":
#24 0.596 mv: cannot stat '': No such file or directory
------
executor failed running [/bin/sh -c mv "$(which runner.sh)" "$(dirname "$(which runner.sh)")/kubekins-runner.sh"]: exit code: 1
make: *** [build] Error 1
```

File: images/prow-tests/Dockerfile [hunk 1]
```diff
@@ -12,42 +12,45 @@
...
-# If we don't run update, the apt-get install below don't work
-RUN apt-get update
+FROM google/cloud-sdk:${GCLOUD_VERSION}-slim AS stable-stuff
 
-# gcloud
-RUN gcloud components update --version=${GCLOUD_VERSION}
+ENV DEBIAN_FRONTEND noninteractive
+
+ARG KUBECTL_VERSION=v1.21.4
+ARG BAZELISK_VERSION=v1.11.0
```
Comment (chizhg) [inline images/prow-tests/Dockerfile:23]: bazel and bazelisk are not needed anymore, so we can safely remove them here and elsewhere.

File: images/prow-tests/Dockerfile [hunk 2]
```diff
@@ -12,42 +12,45 @@
...
     chmod +x /usr/local/bin/kubectl
 
-# Extra tools through apt-get
-RUN apt-get install -y uuid-runtime   # for uuidgen
-RUN apt-get install -y shellcheck     # for shellcheck presubmit
+# bazel
+RUN curl -fsSL "https://github.com/bazelbuild/bazelisk/releases/download/${BAZELISK_VERSION}/bazelisk-linux-amd64" -o /usr/local/bin/bazel && \
+    chmod +x /usr/local/bin/bazel
+
+# Docker including docker-ce, docker-ce-cli, and containerd.io
```
Comment (chizhg) [inline images/prow-tests/Dockerfile:34]: We should need other Docker setups in https://github.com/kubernetes/test-infra/blob/master/images/bootstrap/Dockerfile#L79-L126 as well.

File: images/prow-tests/Dockerfile [hunk 3]
```diff
@@ -100,20 +103,23 @@ ARG KUBETEST2_VERSION=ea3c260668075a29dd2b7f3ba4ddf1fe78bdf898
 ARG KIND_VERSION=v0.11.1
 ARG KO_VERSION=v0.8.2
 ARG PROTOC_GEN_GO_VERSION=v1.26.0
+ARG KIND_VERSION=v0.11.1
```
Comment (chizhg) [inline images/prow-tests/Dockerfile:106]: KIND_VERSION is already defined above (L103).

File: images/prow-tests/Dockerfile [hunk 4]
```diff
@@ -12,42 +12,45 @@
...
+
+# Docker including docker-ce, docker-ce-cli, and containerd.io
+RUN curl -fsSL https://download.docker.com/linux/ubuntu/gpg | apt-key add - && \
+    echo "deb [arch=amd64] https://download.docker.com/linux/debian $(lsb_release -cs) stable" | tee /etc/apt/sources.list.d/docker.list && \
+    apt-get update && \
+    apt-get -y install --no-install-recommends docker-ce="${DOCKER_VERSION}" docker-ce-cli="${DOCKER_VERSION}"
 
 # Setup adoptopenjdk
 ARG JAVA_VERSION=16
-RUN apt-get install -y wget apt-transport-https gnupg
```
Comment (chizhg) [inline images/prow-tests/Dockerfile:38]: What's the reason for deleting this line?

File: images/prow-tests/Dockerfile [hunk 5]
```diff
@@ -12,42 +12,45 @@
...
-RUN wget -qO - https://adoptopenjdk.jfrog.io/adoptopenjdk/api/gpg/key/public | apt-key add -
+RUN curl -sf https://adoptopenjdk.jfrog.io/adoptopenjdk/api/gpg/key/public | apt-key add -
```
Comment (chizhg) [inline images/prow-tests/Dockerfile:42]: What's the reason for changing `wget` to `curl`?

File: images/prow-tests/Dockerfile [hunk 6]
```diff
@@ -100,20 +103,23 @@ ARG KUBETEST2_VERSION=ea3c260668075a29dd2b7f3ba4ddf1fe78bdf898
...
-RUN go get github.com/raviqqe/liche@v0.0.0-20200229003944-f57a5d1c5be4  # stable liche version for checking md links
-RUN go get sigs.k8s.io/kind@${KIND_VERSION}
-RUN go get sigs.k8s.io/kubetest2@${KUBETEST2_VERSION}
-RUN go get sigs.k8s.io/kubetest2/kubetest2-gke@${KUBETEST2_VERSION}
-RUN go get sigs.k8s.io/kubetest2/kubetest2-kind@${KUBETEST2_VERSION}
-RUN go get sigs.k8s.io/kubetest2/kubetest2-tester-exec@${KUBETEST2_VERSION}
-RUN go get google.golang.org/protobuf/cmd/protoc-gen-go@${PROTOC_GEN_GO_VERSION}
-RUN go get github.com/google/go-licenses
+RUN go install github.com/google/ko@${KO_VERSION}
+RUN go get github.com/boz/kail/cmd/kail@v0.15.0
```
Comment (chizhg) [inline images/prow-tests/Dockerfile:111]: Since you are here, would you mind also moving these remaining hardcoded versions to be ARGs?

**Review by chizhg — COMMENTED** [review]
I cannot tell whether we are missing any tools, since the image is now used by all knative and knative-sandbox projects. We can test the image in staging for a while (there is a staging process for rolling out the new image https://github.com/knative/test-infra/tree/main/images/prow-tests#testing-and-deploying-images), and fix them if we see any issues.

**Comment by chizhg** [comment]
> Also, how do you want to deal with the missing file in runner.sh?

We should copy most of the code from https://github.com/kubernetes/test-infra/blob/master/images/bootstrap/runner.sh. It can be directly copied to `images/prow-tests/runner.sh` so we can drop the `RUN mv "$(which runner.sh)" "$(dirname "$(which runner.sh)")/kubekins-runner.sh"` line from the Docker file.

File: images/prow-tests/Dockerfile (hunk 6 above)
Comment (chizhg) [reply images/prow-tests/Dockerfile:111]: Also it'll be clearer if all the ARG lines are at the beginning of this file.

File: images/prow-tests/Dockerfile (hunk 4 above)
Comment (upodroid) [reply images/prow-tests/Dockerfile:38]: This is handled in the gcloud image. It is linked at the top for easy reference

File: images/prow-tests/Dockerfile (hunk 5 above)
Comment (upodroid) [reply images/prow-tests/Dockerfile:42]: So I don't have to install wget. curl is already bundled

Comment (chizhg) [reply images/prow-tests/Dockerfile:42]: I did some search and it seems at least serving is using `wget`. So let's keep it?

Comment (upodroid) [reply images/prow-tests/Dockerfile:42]: i'll add it back in

File: images/prow-tests/Dockerfile (hunk 1 above)
Comment (upodroid) [reply images/prow-tests/Dockerfile]: removed

**Comment by upodroid** [comment]
This is ready for review now.

@chizhg @dprotaso

File: images/prow-tests/Dockerfile [hunk 7]
```diff
@@ -100,57 +140,58 @@ ARG KUBETEST2_VERSION=ea3c260668075a29dd2b7f3ba4ddf1fe78bdf898
...
+RUN go install github.com/GoogleCloudPlatform/docker-credential-gcr@${DOCKER_CREDENTIAL_GCR_VERSION}
 
 # We do this instead of `go get` so the tools are locked to the Dockerfile's commit
 COPY . /go/src/knative.dev/test-infra
 
 # Build custom tools in the container
 
-RUN go install knative.dev/test-infra/tools/coverage
-RUN go install knative.dev/test-infra/kntest/cmd/kntest
+RUN go install knative.dev/test-infra/tools/coverage@latest
```
Comment (chizhg) [inline images/prow-tests/Dockerfile:178]: The reason we didn't add the version is explained at L173 - `We do this instead of "go get" so the tools are locked to the Dockerfile's commit`. We also run the image build in presubmit so that makes sure the tool version we install is always consistent. Can we revert this change?

**Review by chizhg — COMMENTED** [review]
Thanks for addressing the comments!

Only one nit then this PR is good to Go.

File: images/prow-tests/Dockerfile (hunk 7 above)
Comment (upodroid) [reply images/prow-tests/Dockerfile:178]: I know but it fails when I don't add latest.

```
Removing intermediate container 4b7cf464eacf
 ---> 86c3a9d15be6
Step 60/69 : COPY . /go/src/knative.dev/test-infra
 ---> a088264a3284
Step 61/69 : RUN go install knative.dev/test-infra/tools/coverage
 ---> Running in e9acd904c7e9
go install: version is required when current directory is not in a module
        Try 'go install knative.dev/test-infra/tools/coverage@latest' to install the latest version
The command '/bin/sh -c go install knative.dev/test-infra/tools/coverage' returned a non-zero code: 1
ERROR
ERROR: build step 0 "gcr.io/cloud-builders/docker" failed: step exited with non-zero status: 1

```

Comment (chizhg) [reply images/prow-tests/Dockerfile:178]: Changing it to `RUN cd /go/src/knative.dev/test-infra && go install ./tools/coverage` will work. Ditto for other commands.

**Comment by upodroid** [comment]
/hold

buster has very old git version, reworking to start from bullseye

**Comment by upodroid** [comment]
/hold cancel

This is ready to merge now

**Comment by chizhg** [comment]
/lgtm
/approve

I'll monitor the beta Prow jobs in https://testgrid.k8s.io/r/knative-own-testgrid/beta-prow-tests#Summary before rolling it out.
